build-n-publish.sh

```diff
@@ -1,28 +1,205 @@
+set -u
 
 # Lines added to get the script running in the script path shell context
 # reference: http://www.ostricher.com/2014/10/the-right-way-to-get-the-directory-of-a-bash-script/
 cd "$(dirname "$0")"
 
-tag="${pkgName}-${pkgVersion}-$(date +%Y%m%d%H%M%S)"
+# To avoid issues with MINGW and Git Bash, see:
+# https://github.com/docker/toolbox/issues/673
+# https://gist.github.com/borekb/cb1536a3685ca6fc0ad9a028e6a959e3
+export MSYS_NO_PATHCONV=1
+export MSYS2_ARG_CONV_EXCL="*"
+
+if [ -n "${version}" ]
+then
+  versionBuild=${commit}
+  # Ugly code to deal with versions
+  # Input:
+  #   version=v12.34.5
+  #   versionBuild=94f85efb9c
+  #   versionPre=0pr
+  # Output:
+  #   versionMayor=pre-v12
+  #   versionMayorMinor=pre-v12.34
+  #   versionMayorMinorPatch=pre-v12.34.5
+  #   versionMayorMinorPatchPre=pre-v12.34.5-0pr
+  #   versionFull=pre-v12.34.5-0pr+94f85efb9c
+  #   versionFullForTag=pre-v12.34.5-0pr_94f85efb9c
+  # region Ugly code to deal with versions
+
+  versionFull=${version}
+
+  if [ -n "${versionPre}" ]
+  then
+    versionFull=${versionFull}-${versionPre}
+  fi
+
+  if [ -n "${versionBuild}" ]
+  then
+    versionFull=${versionFull}+${versionBuild}
+  fi
+
+  # https://semver.org/spec/v2.0.0.html#backusnaur-form-grammar-for-valid-semver-versions
+  # <valid semver> ::= <version core>
+  #                  | <version core> "-" <pre-release>
+  #                  | <version core> "+" <build>
+  #                  | <version core> "-" <pre-release> "+" <build>
+  #
+  # <version core> ::= <major> "." <minor> "." <patch>
+  versionBuild="$(echo "${versionFull}"+ | cut -d'+' -f2)"
+  versionMayorMinorPatchPre="$(echo "${versionFull}" | cut -d'+' -f1)" # v0.0.0-xxx (ignoring `+` if exists)
+  versionPre="$(echo "${versionMayorMinorPatchPre}"- | cut -d'-' -f2)"
+  versionMayorMinorPatch="$(echo "${versionMayorMinorPatchPre}" | cut -d'-' -f1)" # v0.0.0 (ignoring `-` if exists)
+  versionMayor="$(echo "${versionMayorMinorPatch}" | cut -d'.' -f1)" # v0
+  versionMinor="$(echo "${versionMayorMinorPatch}" | cut -d'.' -f2)"
+  versionMayorMinor="${versionMayor}.${versionMinor}" # v0.0
+  # by the moment we do not need it, versionPatch only for demo purposes
+  # versionPatch="$(echo "${versionMayorMinorPatch}" | cut -d'.' -f3)"
+
+  if [ -z "${versionBuild}" ]
+  then
+    canonicalTag=${versionMayorMinorPatchPre}
+  else
+    # because `+` is not accepted in tag names
+    canonicalTag=${versionMayorMinorPatchPre}_${versionBuild}
+  fi
+
+  if [ -n "${versionPre}" ]
+  then
+    preReleasePrefix="pre-"
+    versionMayor=${preReleasePrefix}${versionMayor}
+    versionMayorMinor=${preReleasePrefix}${versionMayorMinor}
+    versionMayorMinorPatch=${preReleasePrefix}${versionMayorMinorPatch}
+    versionMayorMinorPatchPre=${preReleasePrefix}${versionMayorMinorPatchPre}
+    versionFull=${preReleasePrefix}${versionFull}
+    canonicalTag=${preReleasePrefix}${canonicalTag}
+  fi
+  # endregion Ugly code to deal with versions
```

Comment:
This is to parse the version number into components.

This code is really ugly but it has been tested in many repositories.

Examples:

## Final version

### Input:

* version=v12.34.5
* versionBuild=94f85efb9c

### Output:

* versionMayor=v12
* versionMayorMinor=v12.34
* versionMayorMinorPatch=v12.34.5
* versionMayorMinorPatchPre=v12.34.5
* versionFull=v12.34.5+94f85efb9c
* versionFullForTag=v12.34.5_94f85efb9c


## pre-release version

### Input:

* version=v12.34.5
* versionBuild=94f85efb9c
* versionPre=beta1

### Output:

* versionMayor=pre-v12
* versionMayorMinor=pre-v12.34
* versionMayorMinorPatch=pre-v12.34.5
* versionMayorMinorPatchPre=pre-v12.34.5-beta1
* versionFull=pre-v12.34.5-beta1+94f85efb9c
* versionFullForTag=pre-v12.34.5-beta1_94f85efb9c